src/Compilers/Core/Rebuild/CompilationOptionsReader.cs

```diff
@@ -271,35 +284,75 @@ public ImmutableArray<MetadataReferenceInfo> GetMetadataReferences()
             return result;
         }
 
-        public ImmutableArray<SourceFileInfo> GetSourceFileInfos(Encoding encoding)
+        public (ImmutableArray<SyntaxTree> SyntaxTrees, ImmutableArray<MetadataReference> MetadataReferences) ResolveArtifacts(
+            IRebuildArtifactResolver resolver,
+            Func<string, SourceText, SyntaxTree> createSyntaxTreeFunc)
         {
-            var sourceFileCount = int.Parse(
-                GetMetadataCompilationOptions()
-                    .GetUniqueOption(CompilationOptionNames.SourceFileCount));
+            var syntaxTrees = ResolveSyntaxTrees();
+            var metadataReferences = ResolveMetadataReferences();
+            return (syntaxTrees, metadataReferences);
 
-            var builder = ImmutableArray.CreateBuilder<SourceFileInfo>(sourceFileCount);
-            foreach (var documentHandle in PdbReader.Documents.Take(sourceFileCount))
+            ImmutableArray<SyntaxTree> ResolveSyntaxTrees()
             {
-                var document = PdbReader.GetDocument(documentHandle);
-                var name = PdbReader.GetString(document.Name);
-
-                var hashAlgorithmGuid = PdbReader.GetGuid(document.HashAlgorithm);
-                var hashAlgorithm =
-                    hashAlgorithmGuid == HashAlgorithmSha1 ? SourceHashAlgorithm.Sha1
-                    : hashAlgorithmGuid == HashAlgorithmSha256 ? SourceHashAlgorithm.Sha256
-                    : SourceHashAlgorithm.None;
+                var sourceFileCount = GetSourceFileCount();
+                var builder = ImmutableArray.CreateBuilder<SyntaxTree>(sourceFileCount);
+                foreach (var tuple in GetSourceTextInfoCore())
+                {
+                    SourceText sourceText;
+                    if (ResolveEmbeddedSource(tuple.DocumentHandle, tuple.SourceTextInfo) is { } embeddedSourceTextInfo)
+                    {
+                        sourceText = embeddedSourceTextInfo.SourceText;
+                    }
+                    else
+                    {
+                        sourceText = resolver.ResolveSourceText(tuple.SourceTextInfo);
```

Comment:
>sourceText = resolver.ResolveSourceText(tuple.SourceTextInfo) [](start = 24, length = 61)

The containing `foreach` appears to be O(N^2).

Reply:
Not quite sure what you mean here, can u elaborate? The `GetSourceTextInfoCore` call is just looping through the source docs and returning a collection, this then just looks over the output and lets the resolver provide the source file as they choose.

Reply:
Sorry I wasn't clear. The `foreach` loops over the `SourceTextInfo` collection and uses the resolver to find the corresponding `SourceText`, but `CompilationRebuildArtifactResolver.ResolveSourceText()` walks through the entire `SyntaxTrees` collection to resolve each `SourceText`.

---
In reply to: [610174881](https://github.com/dotnet/roslyn/pull/52511#discussion_r610174881) [](ancestors = 610174881)

Reply:
Ah yes but `CompilationRebuildArtifactResolver` is a test only implementation. It's not meant for production purposes. The expectation is that production implementations will use the hash / file path for an O(1) lookup.